src/persist-client/src/internal/gc.rs

```diff
@@ -237,45 +238,52 @@ where
-                // through the live_diffs loop because they accumulate in state.
-                for (seqno, key) in state.collections.rollups.iter() {
-                    // SUBTLE: We only guarantee that a rollup exists for the
-                    // first live state. Anything before that is not allowed to
-                    // be used and so is free to be deleted and removed from
-                    // state.
-                    if seqno < &earliest_live_seqno {
-                        deleteable_rollup_blobs.push((*seqno, key.clone()));
-                    } else {
-                        // We iterate in order, may as well short circuit the
-                        // rollup loop.
-                        break;
+            match state.seqno.cmp(&req.new_seqno_since) {
+                Ordering::Less => {
+                    state.collections.trace.map_batches(|b| {
+                        for part in b.parts.iter() {
+                            // It's okay (expected) if the key already exists in
+                            // deleteable_batch_blobs, it may have been present in
+                            // previous versions of state.
+                            deleteable_batch_blobs.insert(part.key.to_owned());
+                        }
+                    });
+                }
+                Ordering::Equal => {
+                    live_diffs += 1;
+                    state.collections.trace.map_batches(|b| {
+                        for part in b.parts.iter() {
+                            // It's okay (expected) if the key doesn't exist in
+                            // deleteable_batch_blobs, it may have been added in
+                            // this version of state.
+                            let _ = deleteable_batch_blobs.remove(&part.key);
+                            live_batch_blobs.insert(part.key.to_owned());
+                        }
+                    });
+                    // We only need to detect deletable rollups in the last iter
+                    // through the live_diffs loop because they accumulate in state.
+                    for (seqno, key) in state.collections.rollups.iter() {
+                        // SUBTLE: We only guarantee that a rollup exists for the
+                        // first live state. Anything before that is not allowed to
+                        // be used and so is free to be deleted and removed from
+                        // state.
+                        if seqno < &earliest_live_seqno {
+                            deleteable_rollup_blobs.push((*seqno, key.clone()));
+                        } else {
+                            // We iterate in order, may as well short circuit the
+                            // rollup loop.
+                            break;
+                        }
                     }
+                    break;
+                }
+                Ordering::Greater => {
+                    break;
```

Comment:
looks like we lost the sanity check assert?

Reply:
Not an accident! See the first commit. (The mutual-exclusivity of the three cases is now guaranteed by the match.)

If you don't find that particularly convincing I'm okay to revert it.

Reply:
oh! the assert wasn't there for mutual exclusivity. it was essentially verifying that the states entering the loop are sorted by seqno

...maybe not so interesting of an assert when I put it that way, so i'm happy with whatever here